lib/graphql/batch/execution_strategy.rb

```diff
@@ -1,57 +1,62 @@
 module GraphQL::Batch
   class ExecutionStrategy < GraphQL::Query::SerialExecution
-    attr_reader :batched_queries
-
-    def initialize
-      @batched_queries = Hash.new{ |hash, key| hash[key] = [] }
-    end
-
     class OperationResolution < GraphQL::Query::SerialExecution::OperationResolution
       def result
-        result = super
-        until execution_strategy.batched_queries.empty?
-          queries = execution_strategy.batched_queries.shift.last
-          queries.first.class.execute(queries)
-        end
-        result
+        GraphQL::Batch.wait(super)
+      ensure
+        GraphQL::Batch::Executor.current.clear
       end
     end
 
     class SelectionResolution < GraphQL::Query::SerialExecution::SelectionResolution
       def result
-        result_hash = super
-        result_hash.each do |key, value|
-          if value.is_a?(FieldResolution)
-            value.result_hash = result_hash
+        lazy_resolve_object(super)
+      end
+
+      private
+
+      def lazy_resolve_object(obj)
+        queries = nil
+        case obj
+        when Array
+          array = obj
+          array.each_with_index do |value, i|
+            if query = lazy_key_assign(array, i, value)
+              (queries ||= []) << query
```

Comment:
this is gross, do the unoptimized version unless we have benchmarks showing the extra object allocations are actually a problem